unpack-trees.c

```diff
@@ -1386,12 +1395,13 @@ static int unpack_callback(int n, unsigned long mask, unsigned long dirmask, str
 
 	/* Are we supposed to look at the index too? */
 	if (o->merge) {
+		int hint = -1;
 		while (1) {
 			int cmp;
 			struct cache_entry *ce;
 
 			if (o->diff_index_cached)
-				ce = next_cache_entry(o);
+				ce = next_cache_entry(o, &hint);
```

Comment:
Now that I look closely at your implementation, this old code had _quadratic_ complexity! That helps me understand why having ~70,000 index entries still took 5 seconds to iterate.

I suppose I was expecting `hint` to be initialized with some information other than `-1`, but this seems to be a good idea regardless of any issues with the sparse index.